omnigibson/scene_graphs/graph_builder.py

```diff
@@ -0,0 +1,272 @@
+
+        return desired_frame_to_world, world_to_desired_frame
+
+    def start(self, scene):
+        assert self.G is None, "Cannot start graph builder multiple times."
+
+        robot = get_robot(scene)
+        self.G = nx.DiGraph() if self.merge_parallel_edges else nx.MultiDiGraph()
+
+        desired_frame_to_world, world_to_desired_frame = self._get_desired_frame(robot)
+        robot_pose = world_to_desired_frame @ get_robot_to_world_transform(robot)
+        robot_bbox_pose, robot_bbox_extent = get_robot_bbox(robot)
+        robot_bbox_pose = world_to_desired_frame @ robot_bbox_pose
+        self.G.add_node(
+            robot, pose=robot_pose, bbox_pose=robot_bbox_pose, bbox_extent=robot_bbox_extent, states={}
+        )
+        self.last_desired_frame_to_world = desired_frame_to_world
+
+    def step(self, scene):
+        assert self.G is not None, "Cannot step graph builder before starting it."
+
+        # Prepare the necessary transformations.
+        robot = get_robot(scene)
+        desired_frame_to_world, world_to_desired_frame = self._get_desired_frame(robot)
+
+        # Update the position of everything that's already in the scene by using our relative position to last frame.
+        old_desired_to_new_desired = world_to_desired_frame @ self.last_desired_frame_to_world
+        for obj in self.G.nodes:
+            self.G.nodes[obj]["pose"] = old_desired_to_new_desired @ self.G.nodes[obj]["pose"]
+            self.G.nodes[obj]["bbox_pose"] = old_desired_to_new_desired @ self.G.nodes[obj]["bbox_pose"]
+
+        # Update the robot's pose. We don't want to accumulate errors because of the repeated transforms.
+        self.G.nodes[robot]["pose"] = world_to_desired_frame @ get_robot_to_world_transform(robot)
+        robot_bbox_pose, robot_bbox_extent = get_robot_bbox(robot)
+        robot_bbox_pose = world_to_desired_frame @ robot_bbox_pose
+        self.G.nodes[robot]["bbox_pose"] = robot_bbox_pose
+        self.G.nodes[robot]["bbox_extent"] = robot_bbox_extent
+
+        # Go through the objects in FOV of the robot.
+        objs_to_add = set(scene.objects)
+        if not self.full_obs:
+            # TODO: This probably does not work
+            # If we're not in full observability mode, only pick the objects in FOV of robot.
+            bids_in_fov = robot.states[object_states.ObjectsInFOVOfRobot].get_value()
+            objs_in_fov = set(
+                scene.objects_by_id[bid]
+                for bid in bids_in_fov
+                if bid in scene.objects_by_id
+            )
+            objs_to_add &= objs_in_fov
+
+        for obj in objs_to_add:
+            # Add the object if not already in the graph
+            if obj not in self.G.nodes:
+                self.G.add_node(obj)
+
+            # Get the relative position of the object & update it (reducing accumulated errors)
+            self.G.nodes[obj]["pose"] = world_to_desired_frame @ T.pose2mat(obj.get_position_orientation())
```

Comment:
I don't understand -- if we do that again here, what's the point of doing it earlier? sorry, i might just be dumb LOL